src/native/external/libunwind/include/libunwind-aarch64.h

```diff
@@ -34,9 +34,13 @@ extern "C" {
 
 #include <inttypes.h>
 #include <stddef.h>
-#include <ucontext.h>
 #include <stdalign.h>
 #include <stdint.h>
+#ifdef __APPLE__
+#include <sys/ucontext.h>
+#else
+#include <ucontext.h>
+#endif
```

Comment:
Please upstream this patch (https://github.com/libunwind/libunwind) and add a line next to https://github.com/dotnet/runtime/blob/a3e1a7267943e8e5b6fe2262dc6ef73452f839c0/src/native/external/libunwind-version.txt#L13

The next update (v1.9) will automatically pick it up from upstream, eliminating the need for manual verification on our end.

Reply:
This is just CI test at this point. This should not even be used on Apple platforms but I think DAC references it for one reason or another. I'd rather resolve that and drop the change in libunwind.

Reply:
Yup, windows and macOS only use HP libunwind for remote unwinding feature. It's a bit unfortunately that llvm-libunwind doesn't provide that and HP libunwind doesn't fully support macOS. Converging to one lib was attempted a few times but there isn't a substantial progress on that front. :(

Reply:
I checked the code and there's actually three unwinding libraries...
- HP libunwind
- LLVM libunwind
- The copied LLVM libunwind logic in remote-uniwind.cpp under `__APPLE__` guard

LLVM libunwind actually does provide remote unwinding but only in the C++ API. That's essentially the API used by UnwindHelpers.cpp in NativeAOT. There's a way to pass custom `AddressSpace` implementation and initial register values. However, using that would be substantial rewrite and I am not sure if we have test coverage for that.

It would probably be possible remove the use of HP libunwind on Apple platforms with relatively minor changes. It's only used for few structs that are not exposed externally and as a fallback that should never be hit (since both compact unwinding and DWARF unwinding is implemented in remote-unwind.cpp). Am I missing something here? Maybe some CrossDAC scenarios?

Reply:
> It would probably be possible to remove the use of HP libunwind on Apple platforms with relatively minor changes.

Sounds good. I think we should go ahead with the switch. Both mac and windows have minimal remote unwind helpers in HP-libunwind which were upstreamed from dotnet/runtime: https://github.com/libunwind/libunwind/tree/master/src/remote

We have the option to build coreclr with llvm-libunwind on linux as well, which has been supported since 2016: https://github.com/dotnet/runtime/commit/5a2478f2ce3d16ca2c03f54608a017d01ef69a12 (not sure if it's been tested recently though). Also, `unw_get_save_loc` API (which was the main reason of using HP-libunwind https://github.com/dotnet/runtime/issues/4193#issuecomment-97586716) was implemented in our in-tree copy of llvm-libunwind since https://github.com/dotnet/corert/commit/c6571c6eb7b99afec7d69cd5f52fb3c5058a2e3c, though we haven’t upstreamed those patches yet. There’s an open issue for this: https://github.com/dotnet/runtime/issues/72655, which is a minor problem; it’s simply something we continue to carry forward during llvm-libunwind version bumps.

The majority of HP-libunwind usage in coreclr is due to `HELPER_METHOD_FRAME` (HMF) and `FCThrow`: https://github.com/dotnet/corert/issues/3784. Work is actively ongoing to remove these, and folks have made good progress: https://github.com/dotnet/runtime/issues/95695. We are down to ~80 instances of HMF from 400+, and  `FCThrow` is down to ~75. Once the HMF removal is completed, we can revisit the remaining uses of HP-libunwind, particularly for remote unwinding.

> CrossDAC

The result of:
`git grep -E 'libunwind|unw_' -- 'src/coreclr' ':!src/coreclr/pal' ':!src/coreclr/nativeaot'`
command suggests that we don’t have direct usages of libunwind outside of the CoreCLR-PAL (I maybe missing something), which means we might be able to switch everything over to llvm-libunwind if we switch PAL to llvm-libunwind. Once HMF is removed, PAL implementation can be reduced to bare minimum without having to work out two libunwind flavors.

AFAICT, the llvm-libunwind project started back in 2013 when Apple contributed their unwinder to LLVM. While the system libunwind on macOS may have diverged somewhat, their C API seems to be the same as llvm-libunwind (which itself uses consistent naming as the original HP-libunwind), with the exception of for three functions we care for: `unw_get_save_loc` (which our llvm-libunwind copy includes), `unw_get_accessors` (which we don’t directly use but cmake introspection is in pal/src/configure.cmake for some reason..), and `unw_get_proc_info_in_range` (for which we have a fallback implementation that matches the one originally ported by us to HP-libunwind upstream, and it works with llvm-libunwind as well).

Reply:
The diff for removing the HP libunwind fallback on Apple platforms seems to be quite small. Let's see if I broke anything on CI. Honestly, not very sure on how to test it. I am not aware of any createdump tests.